### ISSUE TITLE
Weighted learning for decision trees

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#882]
There is a new implementation of decision trees in `src/mlpack/methods/decision_tree/`.  It's meant to eventually replace the decision stump implementation in `src/mlpack/methods/decision_stump/` (among other things).  However, it can't fully replace the decision stump until it supports weighted learning.  Right now, one of the `Train()` methods for decision trees is:

```
template<typename MatType>
void Train(const MatType& data,
           const arma::Row<size_t>& labels,
           const size_t numClasses,
           const size_t minimumLeafSize = 10);
```

but note that this does not take weights with the data.  On the other hand, the decision stump has the following `Train()` method:

```
void Train(const MatType& data,
           const arma::Row<size_t>& labels,
           const arma::rowvec& weights,
           const size_t classes,
           const size_t bucketSize);
```

The `weights` vector specifies the weight that each point in `data` should be given.  So, for instance, if the weight of a point is 0, then that point should not make a difference at all during training.  One example for how this is used is in `GaussianDistribution::Train()`: you can take a look at both overloads and see how the weights (called `probabilities` in that class) are used.

The goal of this ticket is to implement weighted learning for decision trees, and add an overload of `Train()` to `DecisionTree` like the above overload for decision stumps.  In order to do this, the gain functions (`information_gain.hpp`, `gini_gain.hpp`) will need to be overloaded to support weighted data, and the splits (`all_categorical_split.hpp`, `best_binary_numeric_split.hpp`) will also need to be overloaded to support weighted data.  Then finally the extra overload will need to be added to `DecisionTree`, and tests should be added to `src/mlpack/tests/decision_tree_test.cpp` for each of those modifications.

**Comment by MikeLing** [comment]
Hi @rcurtin , I would like to work on this after the **Implement multidimensional discrete distribution**  been merged. :)

**Comment by rcurtin** [comment]
Sure, that means that I should keep on top of that then, let me look now... :)

**Comment by DieHumbleX** [comment]
Hi @rcurtin can you assign me this one? there aren't many moderate level issues left :)

**Comment by MikeLing** [comment]
Hi @DieHumbleX, I'm working on this issue already but haven't add test for it due to write GSoC proposal these days. :)

**Comment by MikeLing** [comment]
Hi @rcurtin, I'm went to the unit test writing stage. However, I'm wondering that what if the length of `weights` doesn't match to the length of `labels`. Like I do pass a weights list(`1.0, 1.1, 2, 2.1, 3`) to `train()` and set `UseWeights` as true. But the label list is (`red, pink, yellow, black, green, green, red...etc`). Do we want it cause an error or just set the weight of reset labels as 0.

**Comment by rcurtin** [comment]
Yep, I see what you mean, I think you should definitely throw an exception in this situation, since there's no way we can still provide meaningful output from the classifier.

**Comment by MikeLing** [comment]
Hi @rcurtin , the compiler looks like failed to found `Evaluate<UseWeights>(labels, numClasses, weights)` function and return an error like https://pastebin.mozilla.org/8981376. But I have no idea why it happened :(    Here is my gist of patch https://gist.github.com/MikeLing/2e1056c873eecbf845c14b318ad4a17f

I know your time is precious and you can't to answer all the questions on github. But I have no idea to solve this error after I search it. Please give me some helps on it. Thank you!

**Comment by rcurtin** [comment]
Hi @MikeLing, did you get this sorted out?  I saw there was some IRC discussion but I did not have time to participate.  I have more time now so I can try and help if you're still having a problem.

**Comment by MikeLing** [comment]
First thing is I'm confusing about this part of code https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp#L29-L32 and how it works. I guess it's about sort all the labels right? Which means if we have label list like "2 1 5 3 8 9 2 0 1" and we may want orderedLabels like "0 1 1 2 2 3 5 8 9" Isn't it? If so, then we may have an issue for it because maybe it failed to do that. Please read this https://gist.github.com/MikeLing/b6c7d0a38ce6079c42e4717582d322cc for more detail. 

The second thing is I don't know which dataset we should use to test weight decision tree. Do we have it in the data folder? Or I need to find one by myself

Thank you!

**Comment by rcurtin** [comment]
Ah, I can explain that.  The goal of `orderedLabels` is not to get an ordered list of labels, but instead to get the labels corresponding to the ordered data.  When we split a node, we have one dimension 'd' of data, and we want to find some split point 'v' so that data with value in dimension 'd' less than the split point 'v' goes to the left node, and data with value in dimension 'd' greater than the split point 'v' goes to the right node.  However, in order to find the best split point 'v', we have to exhaustively search all possible split points and calculate the gain (information gain or Gini gain) of each of these possible splits.  These gain measures depend on the labels that go to the left and right nodes; therefore, we must gather those labels, and this is done by sorting the data in increasing order, and then getting the labels corresponding to that ordering.  This is what `orderedLabels` represents.  You can read more about what I've written here by consulting basically any decision tree paper, like the CART paper or the C4.5 paper by Quinlan.

For the second question, you can see the other datasets I've used for the regular decision tree testing.  I use the vc2 dataset (`SimpleGeneralizationTest`), some random datasets, and a spiral dataset with noisy features (`CategoricalBuildTest`).  You can take a look at those tests, and here are some strategies for testing weights:

 * Take a dataset and train it with no weights, and with all weights set to 1.  The results should be identical.
 * Take a dataset and then append many points that are just random noise.  Weight the random noise very small and the regular points very high, and ensure that the tree gets similar performance on some test data as a tree trained with only the original dataset and no weights.
 * Do the same as above, but set the weights of random noise to 0 and the regular points to 1, and ensure that the tree gives the exact same results as a tree trained without weights.

Those ideas are just for the high-level `DecisionTree` testing, but you should also come up with some similar ideas or extensions of existing tests for `GiniGain` and `InformationGain` and `BestBinaryNumericSplit` and `AllCategoricalSplit`.

Let me know if I can clarify anything.

**Comment by MikeLing** [comment]
#947 Finally got something to ask feedback :)

**Comment by rcurtin** [comment]
Fixed with #947.
